pkg/runtime/bulk_subscriber.go

```diff
@@ -0,0 +1,706 @@
+	}
+
+	spans := make([]trace.Span, len(msg.entries))
+	n := 0
+	for _, cloudEvent := range msg.cloudEvents {
+		if iTraceID, ok := cloudEvent[pubsub.TraceIDField]; ok {
+			if traceID, ok := iTraceID.(string); ok {
+				sc, _ := diag.SpanContextFromW3CString(traceID)
+				spanName := fmt.Sprintf("pubsub/%s", msg.topic)
+
+				// no ops if trace is off
+				var span trace.Span
+				ctx, span = diag.StartInternalCallbackSpan(ctx, spanName, sc, a.globalConfig.Spec.TracingSpec)
+				ctx = diag.SpanContextToGRPCMetadata(ctx, span.SpanContext())
+				spans[n] = span
+				n++
+			} else {
+				log.Warnf("ignored non-string traceid value: %v", iTraceID)
+			}
+		}
+	}
+	spans = spans[:n]
+	defer endSpans(spans)
+	ctx = invokev1.WithCustomGRPCMetadata(ctx, msg.metadata)
+
+	clientV1 := runtimev1pb.NewAppCallbackAlphaClient(a.grpc.AppClient)
+
+	start := time.Now()
+	res, err := clientV1.OnBulkTopicEventAlpha1(ctx, envelope)
+	elapsed := diag.ElapsedSince(start)
+
+	for _, span := range spans {
+		if span != nil {
+			m := diag.ConstructSubscriptionSpanAttributes(envelope.Topic)
+			diag.AddAttributesToSpan(span, m)
+			diag.UpdateSpanStatusFromGRPCError(span, err)
+		}
+	}
+
+	if err != nil {
+		errStatus, hasErrStatus := status.FromError(err)
+		if hasErrStatus && (errStatus.Code() == codes.Unimplemented) {
+			// DROP
+			log.Warnf("non-retriable error returned from app while processing bulk pub/sub event: %s", err)
+			bulkSubDiag.statusWiseDiag[string(pubsub.Drop)] += int64(len(msg.entries))
+			bulkSubDiag.elapsed = elapsed
+			populateBulkSubscribeResponsesWithError(msg.entries, bulkResponses, &entryIDIndexMap, nil)
+			return nil
+		}
+
+		err = errors.Errorf("error returned from app while processing bulk pub/sub event: %s", err)
+		log.Debug(err)
+		bulkSubDiag.statusWiseDiag[string(pubsub.Retry)] += int64(len(msg.entries))
+		bulkSubDiag.elapsed = elapsed
+		populateBulkSubscribeResponsesWithError(msg.entries, bulkResponses, &entryIDIndexMap, err)
+		// on error from application, return error for redelivery of event
+		return nil
+	}
+
+	hasAnyError := false
+	for _, response := range res.GetStatuses() {
+		if entryID, ok := entryIDIndexMap[response.EntryID]; ok {
+			switch response.GetStatus() {
+			case runtimev1pb.TopicEventResponse_SUCCESS: //nolint:nosnakecase
+				// on uninitialized status, this is the case it defaults to as an uninitialized status defaults to 0 which is
+				// success from protobuf definition
+				bulkSubDiag.statusWiseDiag[string(pubsub.Success)] += 1
+				(*bulkResponses)[entryID].EntryID = response.EntryID
+				(*bulkResponses)[entryID].Error = nil
+			case runtimev1pb.TopicEventResponse_RETRY: //nolint:nosnakecase
+				bulkSubDiag.statusWiseDiag[string(pubsub.Retry)] += 1
+				(*bulkResponses)[entryID].EntryID = response.EntryID
+				(*bulkResponses)[entryID].Error = errors.Errorf("RETRY status returned from app while processing pub/sub event for entry id: %v", response.EntryID)
+				hasAnyError = true
```

Comment:
Reviewers: Doesn't return from any point, just captures details, marks required stuff and continues on. Will return only when all message responses are iterated through.